ActionButton.js

```diff
@@ -17,220 +17,159 @@ import {
   DEFAULT_ACTIVE_OPACITY
 } from "./shared";
 
-export default class ActionButton extends Component {
-  constructor(props) {
-    super(props);
+const ActionButton = (props) => {
+  const [resetToken, setResetToken] = useState(props.resetToken);
+  const [active, setActive] = useState(props.active);
+  const anim = useRef(new Animated.Value(props.active ? 1 : 0));
+  const timeout = useRef(null);
+  const mounted = useRef();
 
-    this.state = {
-      resetToken: props.resetToken,
-      active: props.active
-    };
-
-    this.anim = new Animated.Value(props.active ? 1 : 0);
-    this.timeout = null;
-  }
 
-  componentDidMount() {
-    this.mounted = true;
-  }
-
-  componentWillUnmount() {
-    this.mounted = false;
-    clearTimeout(this.timeout);
-  }
+  useEffect(() => {
+    mounted.ref = true;
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.resetToken !== this.state.resetToken) {
-      if (nextProps.active === false && this.state.active === true) {
-        if (this.props.onReset) this.props.onReset();
-        Animated.spring(this.anim, { toValue: 0 }).start();
-        setTimeout(
-          () =>
-            this.setState({ active: false, resetToken: nextProps.resetToken }),
-          250
-        );
-        return;
-      }
+    return () => {
+      mounted.ref = false;
+      timeout.current && clearTimeout(timeout.current);
+    }
+  }, []);
 
-      if (nextProps.active === true && this.state.active === false) {
-        Animated.spring(this.anim, { toValue: 1 }).start();
-        this.setState({ active: true, resetToken: nextProps.resetToken });
-        return;
-      }
+  useEffect(() => {
+    if (props.active) {
+      props.onReset && props.onReset();
 
-      this.setState({
-        resetToken: nextProps.resetToken,
-        active: nextProps.active
-      });
+      Animated.spring(anim.current., { toValue: 0 }).start();
```

Comment:
This throws an error when I tried it out. Couldn't build.

Reply:
Looks like this `anim.current.` is there a couple times, maybe a syntax I'm not familiar with? We're in a typescript react-native project if that helps.